pkg/network/node/ovscontroller.go

```diff
@@ -767,35 +767,51 @@ func (oc *ovsController) ensureTunMAC() error {
 }
 
 func (oc *ovsController) SetNamespaceEgressNormal(vnid uint32) error {
+	_ = oc.ovs.DeleteGroup(vnid)
 	otx := oc.ovs.NewTransaction()
 	otx.DeleteFlows("table=100, reg0=%d", vnid)
 	return otx.Commit()
 }
 
 func (oc *ovsController) SetNamespaceEgressDropped(vnid uint32) error {
+	_ = oc.ovs.DeleteGroup(vnid)
 	otx := oc.ovs.NewTransaction()
 	otx.DeleteFlows("table=100, reg0=%d", vnid)
 	otx.AddFlow("table=100, priority=100, reg0=%d, actions=drop", vnid)
 	return otx.Commit()
 }
 
-func (oc *ovsController) SetNamespaceEgressViaEgressIP(vnid uint32, nodeIP, mark string) error {
+//func (oc *ovsController) SetNamespaceEgressViaEgressIP(vnid uint32, nodeIP, mark string) error {
+func (oc *ovsController) SetNamespaceEgressViaEgressIP(vnid uint32, marks map[string]string) error {
 	otx := oc.ovs.NewTransaction()
 	otx.DeleteFlows("table=100, reg0=%d", vnid)
-	if nodeIP == "" {
-		// Namespace wants egress IP, but no node hosts it, so drop
-		otx.AddFlow("table=100, priority=100, reg0=%d, actions=drop", vnid)
-	} else if nodeIP == oc.localIP {
-		if err := oc.ensureTunMAC(); err != nil {
-			return err
+	_ = oc.ovs.DeleteGroup(vnid)
+
+	var buildBuckets []string
+	for nodeIP, markedEgressIP := range marks {
+		if nodeIP == oc.localIP {
+			if err := oc.ensureTunMAC(); err != nil {
+				return err
+			}
+			buildBuckets = append(buildBuckets, fmt.Sprintf("actions=set_field:%s->eth_dst,set_field:%s->pkt_mark,output:tun0", oc.tunMAC, markedEgressIP))
```

Comment:
If an egress IP is available locally, you should _always_ process it locally.

In particular, the way this is written now, if an egress IP exists on Nodes A and B, then it's possible that Node C might send traffic to Node A for the egress IP, but then Node A will decide to forward it to Node B, and then Node B will forward it back to Node A, and then Node A will emit it locally, etc. The more backup egress IPs you have, the less likely it is the packet will ever actually egress :joy:

Reply:
done